File: src/main/java/com/meowmeowland/rl/QLearning.java
package main.java.com.meowmeowland.rl;

import com.example.sqlgame.environment.SQLGameEnvironment;
import com.example.sqlgame.model.GameState;

import java.util.Random;

/**
 * Basic Q-Learning logic with a Q-table.
 * We'll discretize the states, so we need a way to map (GameState) to an index.
 */
public class QLearning {
    
    // Hyperparameters
    private double alpha;         // Learning rate
    private double gamma;         // Discount factor
    private double epsilon;       // Exploration rate
    private double epsilonDecay;
    private double epsilonMin;
    private int numStates;        // we define a state encoding scheme
    private int numActions;

    private double[][] qTable;
    private Random rand = new Random();

    public QLearning(int numStates, int numActions,
                     double alpha, double gamma, double epsilon,
                     double epsilonDecay, double epsilonMin) {
        this.numStates = numStates;
        this.numActions = numActions;
        this.alpha = alpha;
        this.gamma = gamma;
        this.epsilon = epsilon;
        this.epsilonDecay = epsilonDecay;
        this.epsilonMin = epsilonMin;

        // Initialize Q-table
        qTable = new double[numStates][numActions];
    }

    public int chooseAction(int stateIndex) {
        // Epsilon-greedy
        if (rand.nextDouble() < epsilon) {
            return rand.nextInt(numActions);
        } else {
            return argMax(qTable[stateIndex]);
        }
    }

    public void updateQ(int currentStateIndex, int action, double reward, int nextStateIndex, boolean done) {
        double oldValue = qTable[currentStateIndex][action];
        double nextMax = 0.0;
        if (!done) {
            nextMax = qTable[nextStateIndex][argMax(qTable[nextStateIndex])];
        }
        double newValue = oldValue + alpha * (reward + gamma * nextMax - oldValue);
        qTable[currentStateIndex][action] = newValue;
    }

    public void decayEpsilon() {
        if (epsilon > epsilonMin) {
            epsilon *= epsilonDecay;
        }
    }

    // Helper to encode GameState -> index
    // In a real system, you'd carefully define the dimension ranges
    public int encodeState(GameState s) {
        // Example dimension ranges (for demonstration):
        // role: 2  (0..1)
        // stage: 6 (0..5)
        // skill: 6 (0..5)
        // time: 101(0..100) -> to keep it simple
        // diff: 3  (0..2)

        int roleFactor = s.getRole();           // 0..1
        int stageFactor = s.getStorylineStage(); // 0..5
        int skillFactor = s.getSqlSkill();       // 0..5
        int timeFactor = s.getEngagementTime();  // 0..100
        int diffFactor = s.getPuzzleDifficulty();// 0..2

        // We'll flatten this:
        // index = roleFactor
        //        + stageFactor   * 2
        //        + skillFactor   * 2 * 6
        //        + timeFactor    * 2 * 6 * 6
        //        + diffFactor    * 2 * 6 * 6 * 101
        // It's crucial to keep the ranges consistent with environment constraints.

        int index = roleFactor;
        index += stageFactor * 2; 
        index += skillFactor * 2 * 6;
        index += timeFactor * 2 * 6 * 6;
        index += diffFactor * 2 * 6 * 6 * 101;
        return index;
    }

    private int argMax(double[] arr) {
        int bestIndex = 0;
        double maxVal = arr[0];
        for (int i = 1; i < arr.length; i++) {
            if (arr[i] > maxVal) {
                maxVal = arr[i];
                bestIndex = i;
            }
        }
        return bestIndex;
    }

    public double[][] getQTable() {
        return qTable;
    }
}

File: src/rl/QLearning.ts
/**
 * Manages a Q-table for discrete states and actions.
 * The developer must define how to "encode" a GameState to a single integer index.
 */
export class QLearning {
    private alpha: number;         // learning rate
    private gamma: number;         // discount factor
    private epsilon: number;       // exploration rate
    private epsilonDecay: number;
    private epsilonMin: number;
    private numStates: number;
    private numActions: number;
    private qTable: number[][];
  
    constructor(
      numStates: number,
      numActions: number,
      alpha: number,
      gamma: number,
      epsilon: number,
      epsilonDecay: number,
      epsilonMin: number
    ) {
      this.numStates = numStates;
      this.numActions = numActions;
      this.alpha = alpha;
      this.gamma = gamma;
      this.epsilon = epsilon;
      this.epsilonDecay = epsilonDecay;
      this.epsilonMin = epsilonMin;
  
      this.qTable = Array.from({ length: numStates }, () => Array(numActions).fill(0));
    }
  
    public chooseAction(stateIndex: number): number {
      // Epsilon-greedy
      if (Math.random() < this.epsilon) {
        return Math.floor(Math.random() * this.numActions);
      }
      return this.argMax(this.qTable[stateIndex]);
    }
  
    public updateQ(
      currentStateIndex: number,
      action: number,
      reward: number,
      nextStateIndex: number,
      done: boolean
    ) {
      const oldValue = this.qTable[currentStateIndex][action];
      let nextMax = 0;
      if (!done) {
        const bestNext = this.argMax(this.qTable[nextStateIndex]);
        nextMax = this.qTable[nextStateIndex][bestNext];
      }
      const newValue = oldValue + this.alpha * (reward + this.gamma * nextMax - oldValue);
      this.qTable[currentStateIndex][action] = newValue;
    }
  
    public decayEpsilon(): void {
      if (this.epsilon > this.epsilonMin) {
        this.epsilon *= this.epsilonDecay;
      }
    }
  
    private argMax(arr: number[]): number {
      let bestIndex = 0;
      let bestVal = arr[0];
      for (let i = 1; i < arr.length; i++) {
        if (arr[i] > bestVal) {
          bestVal = arr[i];
          bestIndex = i;
        }
      }
      return bestIndex;
    }
  
    public getQTable(): number[][] {
      return this.qTable;
    }
  }
  